Expose proximity boosting

mayya-sharipova opened the pull request:
Expose DistanceFeatureQuery for geo, date_nanos an date types

Closes #33382

mayya-sharipova committed:
Expose proximity boosting

Expose DistanceFeatureQuery for long, geo and date types

Closes #33382

elasticmachine commented:
Pinging @elastic/es-search

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/docbldesx

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,190 @@
+        }
+    }
+
+    public static Object originFromXContent(XContentParser parser) throws IOException {
```
nit: can be private

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,190 @@
+        try {
+            DistanceFeatureQueryBuilder builder = PARSER.apply(parser, null);
+            return builder;
+        } catch (IllegalArgumentException e) {
```
Why are you catching IAE here? I don't think we usually do that but just bubble up exceptions instead. Maybe it has advantages I'm missing though.

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,190 @@
+            return parser.text();
+        } else if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            return GeoUtils.parseGeoPoint(parser);
+        } else { // if currentToken equals to START_ARRAY)
```
Can you check START_ARRAY explicitely here and otherwise throw an error if we encounter an unexpected token?

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,190 @@
+        }
+    }
+
+    public static Object pivotFromXContent(XContentParser parser) throws IOException {
```
nit: private

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,190 @@
+    public static Object pivotFromXContent(XContentParser parser) throws IOException {
+        if (parser.currentToken() == XContentParser.Token.VALUE_NUMBER) {
+            return parser.longValue();
+        } else { // if currentToken equals to VALUE_STRING
```
same as above, please check the expected token type and throw error if anything unexpected shows up in the token stream.

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,190 @@
+        declareStandardFields(PARSER);
+    }
+
+    public DistanceFeatureQueryBuilder(String field, Object origin, Object pivot) {
```
Since all arguments are required, add a null-check please.

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,190 @@
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(field);
+        out.writeGenericValue(origin);
```
this could move to an inner class if we added them for Origin and Pivot.

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,190 @@
+            long pivotLong = (Long) pivot;
+            return LongPoint.newDistanceFeatureQuery(field, boost, originLong, pivotLong);
+        } else if (fieldType instanceof GeoPointFieldType) {
+            GeoPoint originGeoPoint = (origin instanceof GeoPoint)? (GeoPoint) origin : GeoUtils.parseFromString((String) origin);
```
nit: space between ")?"

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,190 @@
+        if (fieldType instanceof DateFieldType) {
+            long originLong = (origin instanceof Long) ? (Long) origin :
+                ((DateFieldType) fieldType).parseToLong(origin, true, null, null, context);
+            TimeValue val = TimeValue.parseTimeValue((String)pivot, TimeValue.timeValueHours(24),
```
nit: space between "(String)pivot"

cbuescher commented on server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java:
```diff
@@ -0,0 +1,208 @@
+        switch (field) {
+            case GEO_POINT_FIELD_NAME:
+                origin = new GeoPoint(randomDouble(), randomDouble());
+                origin = randomBoolean()? origin : ((GeoPoint) origin).geohash();
```
nit: spaces

cbuescher commented on server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java:
```diff
@@ -0,0 +1,208 @@
+        float boost = queryBuilder.boost;
+        final Query expectedQuery;
+        if (fieldName.equals(GEO_POINT_FIELD_NAME)) {
+            GeoPoint originGeoPoint = (origin instanceof GeoPoint)? (GeoPoint) origin : GeoUtils.parseFromString((String) origin);
```
nit: spaces

cbuescher reviewed (changes requested):
@mayya-sharipova this looks like a really cool and useful feature. I left a couple of comments, most of them minor. I think the biggest suggestion I'd have would be trying to avoid to have untyped (Object) origin and pivot fields and instead create some very simple inner class for holding this information in either one of the allowed types (long, String, Date). This way we should be able to control which values the user can set and prevent accidental misuses early.
I was also wondering if it makes sense to add a few integration tests (something extending ESIntegTestCase) to check if the scoring works as advertised (e.g. boosts are applied correctly, pivot works as expected etc...) but then again I guess this is already covered in the underlying Lucene queries so maybe its not that useful, but wanted to bring it up as a suggestion.
Other than that great feature, looking forward to using it in 7.1.

mayya-sharipova commented on docs/reference/query-dsl/distance-feature-query.asciidoc:
```diff
@@ -0,0 +1,214 @@
+ways to modify the score, this query has the benefit of being able to
+efficiently skip non-competitive hits when
+<<search-uri-request,`track_total_hits`>> is not set to `true`. Speedups may be
+spectacular.
```
@cbuescher I have copied this phrase from [rank_feature query](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/query-dsl-rank-feature-query.html) which is using the same optimizations, and we also have a [blog post](https://www.elastic.co/blog/faster-retrieval-of-top-hits-in-elasticsearch-with-block-max-wand) on this.  This blog post has a link to Lucene benchmarks, but looks like  adding this link to these benchmarks would be excessive here.

jpountz commented on docs/reference/query-dsl/special-queries.asciidoc:
```diff
@@ -28,6 +28,12 @@ the specified document.
+<<query-dsl-distance-feature-query,`distance_feature` query>>::
+
+A query that computes scores based on the dynamically computed distances
+between the origin and documents' long numeric, geo or distance fields.
```
should it mention dates?

jpountz commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,190 @@
+                ((DateFieldType) fieldType).parseToLong(origin, true, null, null, context);
+            TimeValue val = TimeValue.parseTimeValue((String)pivot, TimeValue.timeValueHours(24),
+                DistanceFeatureQueryBuilder.class.getSimpleName() + ".pivot");
+            long pivotLong = val.getMillis();
```
this isn't correct with `date_nanos`.

jpountz reviewed (commented):
This looks good overall. Two thoughts on the PR:
 - I'm wondering that we should only expose it on `date(_nanos)` and `geo_point` fields for now. Lucene exposes it only longs mostly because it doesn't have dedicated fields for dates. As-is I think it feels a bit inconsistent to support this feature on longs but not integers or doubles. Since it's also likely less useful than recency boosting I'm wondering that we should leave it out for now?
 - When I opened the issue I wondered whether it would make more sense to have one query for all data types or one query per data type. Looking at the PR I'm wondering that having one query per data type might make things cleaner. For instance it's a pity that some validation only occurs at the shard level while we could do it at the coordinating node level if we knew whether the query would apply to a geo point or a date field.

jpountz commented:
I'm adding the `release highlight` label as I know many people including @qhoxie expressed interest for efficient recency and geo-distance boosting in the past.

mayya-sharipova commented:
@cbuescher  thanks for such an extensive review - good feedback on the class design
@jpountz  thank for the review, Adrien. What you suggested makes sense, I will redesign this PR to have separate queries and only for  date and geo fields.

@jpountz  @cbuescher I am wondering what should be the name for these new queries. I have two options:
1) `distance_feature_date` and ` distance_feature_geo` - option in line with another similar query - `rank_feature` query
2) `recency_boosting` (for date) and `proximity_boosting` (for geo)  -  option to easier for users to understand what these queries are doing

cbuescher commented:
> I wondered whether it would make more sense to have one query for all data types or one query per data type. 
> Having one query per data type might make things cleaner. For instance it's a pity that some validation only occurs at the shard level while we could do it at the coordinating node level

I agree that being able to validate at the coordinating node would be great. Separating the query however comes with some burden (two query builders, test, documentations etc...) that might be getting even bigger if we choose to add different proximity types later. Maybe it would be possible to stay with one `distance_feature` query but introduce a mandatory `type` argument (e.g. `date`, `geo` etc...) that we then could use for validation on the coordinating node? @jpountz wdyt?

jpountz commented:
@cbuescher What would the API look like, are you thinking of something like below

```
"distance_feature": {
  "type": "geo",
  "field": "location",
  "origin": <origin>,
  "pivot": <pivot>,
  "boost" : <boost>
}
```

That said, we have a single query for the `range` query even though it suffers for the same issue, so maybe we should keep the current approach for consistency.

cbuescher commented:
@jpountz yes, thats what I had in mind. This way I think we can check the input parameters for consistency already on the coordinating node when parsing the query, e.g. check that dates can be parsed etc...

mayya-sharipova commented:
@jpountz @cbuescher  thanks for suggestions. I will go then with the format:
```
"distance_feature": {
  "type": "geo",
  "field": "location",
  "origin": <origin>,
  "pivot": <pivot>,
  "boost" : <boost>
}
```
For the license, it looks like we are sticking with OSS.

jimczi commented:
> That said, we have a single query for the range query even though it suffers for the same issue, so maybe we should keep the current approach for consistency.

And for simplicity ;), exposing the `type` argument just to allow parsing on the coordinating node seems overkill to me and I am not sure that `type` would be enough to handle `date` format, ...
I am not too concerned by the fact that parsing errors will be detected at the shard level, most of our queries that require parsing have the same limitations so IMO we shouldn't add extra arguments that are useful only internally.

cbuescher commented:
> I am not too concerned by the fact that parsing errors will be detected at the shard level, most of our queries that require parsing have the same limitations so IMO we shouldn't add extra arguments that are useful only internally.

I'm okay with either way that doesn't involve adding several differen query builders etc...

jpountz commented:
Thanks @jimczi for chiming in, agreed.

mayya-sharipova committed:
Address feedback:

- Modify distance_feature query to be on date, date_nanos and geo
    fields only. Exclude long field
- Make `origin` as an inner class

mayya-sharipova commented on server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java:
```diff
@@ -545,6 +545,16 @@ private static GeoPoint parseGeoHash(GeoPoint point, String geohash, EffectivePo
         }
     }
 
+    public static GeoPoint parseFromString(String val){
```
@cbuescher I have added the documentation for `parseFromString`. I have not added any checks for invalid input, as these checks are done withing interal functions that `parseFromString` calls.

mayya-sharipova commented:
@cbuescher @jpountz  Thanks for your review. This is ready for another round of review. Changes made:

- Modify distance_feature query to be on date, date_nanos and geo ields only. Exclude long field
- Make `origin` as an inner class

mayya-sharipova committed:
Merge branch 'master' into expose-distance-feature-query

mayya-sharipova commented:
@elasticmachine elasticsearch-ci/1

mayya-sharipova commented:
@elasticmachine  test this please

jpountz commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,215 @@
+                return LongPoint.newDistanceFeatureQuery(field, boost, originLong, pivotVal.getNanos());
+            }
+        } else if (fieldType instanceof GeoPointFieldType) {
+            GeoPoint originGeoPoint = (originObj instanceof GeoPoint)? (GeoPoint) originObj : GeoUtils.parseFromString((String) originObj);
```
the cast from String might raise a ClassCastException?

jpountz commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,215 @@
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        MappedFieldType fieldType = context.fieldMapper(field);
+        if (fieldType == null) {
+            throw new IllegalArgumentException("Can't run [" + NAME + "] query on unmapped fields!");
```
our general policy is to be lenient when it comes to unmapped fields, to make cross-index search easier. I'd return a `MatchNoDocsQuery` instead.

jpountz commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,215 @@
+        Object originObj = origin.origin();
+        if (fieldType instanceof DateFieldType) {
+            long originLong = (originObj instanceof Long) ? (Long) originObj :
+                ((DateFieldType) fieldType).parseToLong(originObj, true, null, null, context);
```
I don't like taking longs as-is, we should always go through `parseToLong` imo. I'd like to avoid that users need to be aware of the internal resolution of the field.

mayya-sharipova replied:
@jpountz  `parseToLong` does NOT work for date_nanos field when parsing a huge long value (an expected value for `date_nanos`). 
`DateFieldType::parseToLong`  uses `JavaDateMathParser` (instead of desired `JodaDateMathParser`) even if a  date type is `date_nanos`. Or should I pass a `forcedDateParser` to `parseToLong` depending on the fieldType's resolution?

mayya-sharipova replied:
@jpountz  Please disregard my previous comment here. I assumed that `date_nanos` except a long value as nanoseconds-since-the-epoch, but actually the long value should be  milliseconds-since-the-epoch.   
I have made the changes as you suggested.

jpountz commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,215 @@
+            return LatLonPoint.newDistanceFeatureQuery(field, boost, originGeoPoint.lat(), originGeoPoint.lon(), pivotDouble);
+        }
+        throw new IllegalArgumentException(
+            "Illegal data type! ["+ NAME + "] query can only be run on a date, date_nanos or geo_point field type!");
```
can you add the type of the field to the error message?

jpountz commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,215 @@
+                this.origin = origin;
+            } else {
+                throw new IllegalArgumentException("Illegal type for [origin]! Must be of type [long] or [string] for " +
+                    "date and date_nanos origins," + "[geo_point] or [string] for geo_point origins!");
```
can you add the class of the object to the error message?

jpountz reviewed (changes requested):
I left some comments.

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,215 @@
+
+    public static class Origin {
+        private final Object origin;
+        public Origin(Object origin) {
```
nit: add a newline befor the constructor

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,215 @@
+            "Illegal data type! ["+ NAME + "] query can only be run on a date, date_nanos or geo_point field type!");
+    }
+
+    public String fieldName() {
```
can be package private I think

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,215 @@
+        return field;
+    }
+
+    public Origin origin() {
```
can be package private I think

cbuescher commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,215 @@
+        return origin;
+    }
+
+    public String pivot() {
```
can be package private I think

cbuescher reviewed (approved):
Left some more minor comments, but nothing huge that needs another review from my side I think.

mayya-sharipova committed:
Address Adrien and Christophs' comments

mayya-sharipova committed:
Merge branch 'master' into expose-distance-feature-query

mayya-sharipova committed:
Revert back origin

mayya-sharipova committed:
Change to milliseconds for date_nanos

mayya-sharipova commented:
@cbuescher @jpountz  Thanks for another round of the  review.  I have addressed all your comments in the last commits.

mayya-sharipova commented:
@elasticmachine  run elasticsearch-ci/packaging-sample

mayya-sharipova committed:
Merge branch 'master' into expose-distance-feature-query

jpountz commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,232 @@
+        Object originObj = origin.origin();
+        if (fieldType instanceof DateFieldType) {
+            long originLong = ((DateFieldType) fieldType).parseToLong(originObj, true, null, null, context);
+            TimeValue pivotVal = TimeValue.parseTimeValue(pivot, TimeValue.timeValueHours(24),
```
It's a bit confusing to pass a default value here given that the origin can't be null.

jpountz commented on server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java:
```diff
@@ -0,0 +1,232 @@
+
+        public Origin(GeoPoint origin) {
+            this.origin = origin;
+        }
```
let's reject nulls in above constructors

jpountz commented on server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java:
```diff
@@ -0,0 +1,237 @@
+                pivot = randomTimeValue(1, 1000, "d", "h", "ms", "s", "m");
+                break;
+            default: // DATE_NANOS_FIELD_NAME
+                Instant randomDateNanos = Instant.now().minus(Duration.ofNanos(randomLongBetween(0, 100_000_000)));
```
Let's avoid using now to make failures reproducible?

jpountz commented on server/src/test/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilderTests.java:
```diff
@@ -0,0 +1,237 @@
+                pivot = randomFrom(DistanceUnit.values()).toString(randomDouble());
+                break;
+            case DATE_FIELD_NAME:
+                long randomDateMills = System.currentTimeMillis() - randomLongBetween(0, 1_000_000);
```
let's avoid using currentTimeMillis to make things reproducible?

jpountz reviewed (approved):
I left some minor comments. Other than that LGTM.

mayya-sharipova committed:
Address Adrien's comments

mayya-sharipova committed:
Correct checkstyle
